ntp-client/src/Network/NTP/Client.hs

```diff
@@ -71,248 +68,308 @@ data NtpStatus =
-
-instance Exception NoHostResolved
-
-updateStatus :: Tracer IO NtpTrace -> NtpState -> IO ()
-updateStatus tracer cli = do
-    offsets <- readTVarIO (ncState cli)
-    status <- case offsets of
-        [] -> do
-           traceWith tracer NtpTraceUpdateStatusNoResponses
-           return NtpSyncUnavailable
-        l -> do
-           let o = ntpSelection (ncSettings cli) $ NE.fromList l
-           traceWith tracer $ NtpTraceUpdateStatusClockOffset $ getNtpOffset o
-           return $ NtpDrift o
-    atomically $ writeTVar (ncStatus cli) status
-
--- Every `ntpPollDelay` we send a request to the list of `ntpServers`.  Before
--- sending a request, we put `NtpSyncPending` to `ncState`.  After sending
--- all requests we wait until either all servers responded or
--- `ntpResponseTimeout` passesed.  If at least one server responded
--- `handleCollectedResponses` will update `ncStatus` in `NtpClient` with a new
--- drift.
-sendLoop :: Tracer IO NtpTrace -> NtpState -> [Addresses] -> IO ()
-sendLoop tracer cli addrs = forever $ do
-    let respTimeout = ntpResponseTimeout (ncSettings cli)
-    let poll        = ntpPollDelay (ncSettings cli)
-
-    -- send packets and wait until end of poll delay
-    sock <- readTVarIO $ ncSockets cli
-    pack <- mkNtpPacket
-    sendPacket tracer sock pack addrs
-
-    _ <- timeout respTimeout waitForResponses
-    updateStatus tracer cli
-    -- after @'updateStatus'@ @'ntpStatus'@ is guaranteed to be
-    -- different from @'NtpSyncPending'@, now we can wait until it was
-    -- changed back to @'NtpSyncPending'@ to force a request.
-    _ <- timeout poll waitForRequest
-
-    -- reset state & status before next loop
-    atomically $ writeTVar (ncState cli) []
-    atomically $ writeTVar (ncStatus cli) NtpSyncPending
 
+setNtpPort :: SockAddr ->  SockAddr
+setNtpPort addr = case addr of
+    (SockAddrInet  _ host)            -> SockAddrInet  ntpPort host
+    (SockAddrInet6 _ flow host scope) -> SockAddrInet6 ntpPort flow host scope
+    sockAddr                   -> sockAddr
+  where
+    ntpPort :: PortNumber
+    ntpPort = 123
+
+createAndBindSock
+    :: Tracer IO NtpTrace
+    -> AddrInfo
+    -> AddrInfo
+    -> IO (Either IOError Socket)
+createAndBindSock tracer addr someDest = do
+    (tryIOError $ Socket.socket (addrFamily addr) Datagram Socket.defaultProtocol) >>= \case
+        Left err -> return $ Left err
+        Right sock -> (tryIOError $ trySock sock) >>= \case
+            Right () -> do
+                traceWith tracer $ NtpTraceSocketCreated (show $ addrFamily addr) (show $ addrAddress addr)
+                return $ Right sock
+            Left err -> do
+                Socket.close sock
+                return $ Left err
+    where
+        trySock s = do
+            Socket.setSocketOption s ReuseAddr 1
+            Socket.bind s (addrAddress addr)
+            -- Send testPacket ?
+            void $ Socket.ByteString.sendTo s "test" (setNtpPort $ Socket.addrAddress someDest)
```

Comment:
I want to catch the case that I open a socket, but the destination is not reachable.
This is exactly what happens when I try it with my network setup.
I can open an IPv6 socket but I cannot send on it, because there are no outgoing routes
for the destination.

Reply:
This will be reviled eventually by sending ntp packet.  This is non-recoverable error, so it's fine to just propagate the error to the application that is using the `ntp-client` to decide what to do with it.

Reply:
No it is more complicated:
If there is an error in the `runQueryLoop`, I terminate `oneshotClient` pause for some delay
and start from a clean state.
If there is an error while setting up IPv4 or IPv6 I start the `runQueryLoop` with whatever protocol is available.
Before I start the `runQueryLoop` with either socket I must ensure that I can send at least one packet.
I need to trigger the error while setting up the sockets, because after the setup the error will propagate differently

Reply:
During setup the error will cause the protocol version to be ignored.
and after setup an error will cause a restart